umap/layouts.py

```diff
@@ -116,70 +122,185 @@ def optimize_layout_euclidean(
     epoch_of_next_negative_sample = epochs_per_negative_sample.copy()
     epoch_of_next_sample = epochs_per_sample.copy()
 
-    for n in range(n_epochs):
-        for i in range(epochs_per_sample.shape[0]):
-            if epoch_of_next_sample[i] <= n:
-                j = head[i]
-                k = tail[i]
+    N = epochs_per_sample.shape[0]
+    effective_n_jobs = effective_n_jobs_with_context(n_jobs)
+    chunk_size = int(math.ceil(N / effective_n_jobs))
+
+    with joblib.Parallel(prefer="threads", n_jobs=n_jobs) as parallel:
+        for n in range(n_epochs):
+
+            _optimize_layout_euclidean_single_epoch_joblib(
+                head_embedding,
+                tail_embedding,
+                head,
+                tail,
+                n_vertices,
+                epochs_per_sample,
+                a,
+                b,
+                rng_state,
+                gamma,
+                dim,
+                move_other,
+                alpha,
+                epochs_per_negative_sample,
+                epoch_of_next_negative_sample,
+                epoch_of_next_sample,
+                n,
+                N,
+                effective_n_jobs,
+                chunk_size,
+                parallel,
+            )
+
+            alpha = initial_alpha * (1.0 - (float(n) / float(n_epochs)))
+
+            if verbose and n % int(n_epochs / 10) == 0:
+                print("\tcompleted ", n, " / ", n_epochs, "epochs")
+
+        return head_embedding
+
+
+def _optimize_layout_euclidean_single_epoch_joblib(
+    head_embedding,
+    tail_embedding,
+    head,
+    tail,
+    n_vertices,
+    epochs_per_sample,
+    a,
+    b,
+    rng_state,
+    gamma,
+    dim,
+    move_other,
+    alpha,
+    epochs_per_negative_sample,
+    epoch_of_next_negative_sample,
+    epoch_of_next_sample,
+    n,
+    N,
+    n_tasks,
+    chunk_size,
+    parallel,
+):
+    def inner_fn(index):
+        local_rng_state = rng_state.copy()
+        seed(local_rng_state, index)
+        _optimize_layout_euclidean_single_epoch_subrange(
+            head_embedding,
+            tail_embedding,
+            head,
+            tail,
+            n_vertices,
+            epochs_per_sample,
+            a,
+            b,
+            local_rng_state,
+            gamma,
+            dim,
+            move_other,
+            alpha,
+            epochs_per_negative_sample,
+            epoch_of_next_negative_sample,
+            epoch_of_next_sample,
+            n,
+            chunk_size * index,
+            min(chunk_size * (index + 1), N),
+        )
+
+    parallel(parallel_calls(inner_fn, n_tasks))
+
+
+@numba.njit(fastmath=True, nogil=True)
```

Comment:
Your question makes a lot of sense. The answer is to do with random number seeding and reproducibility: basically Numba parallel doesn't allow controlled seeding across threads so it's not possible to get a reproducible result. With this code (using joblib), if you don't set a seed it will use all cores, otherwise it will use a single core and the result will be deterministic. See discussion in #231 for why that is not achievable with Numba parallel.

Reply:
You could just toggle the decorator.
```python
import numba
import numpy as np

def searchsorted(a, v):
    indices = np.empty(v.shape, dtype=np.int64)
    for i in numba.prange(v.shape[0]):
        indices[i] = np.searchsorted(a[i], v[i])
    return indices

# No numba, no parallel
searchsorted(np.array([[4,3,2,5]]), np.array([[1]]))

# Now uses the numba implementation
parallel_searchsorted = numba.njit(searchsorted)

parallel_searchsorted(np.array([[4,3,2,5]]), np.array([[1]]))
```

Reply:
This way you can also programmatically just toggle parallel on or off by passing it to njit.
See https://github.com/numba/numba/blob/master/numba/decorators.py#L41